### ISSUE TITLE
Improving rendering performance for large number of rows

### ISSUE THREAD

**Issue by sehilyi** [hms-dbmi/cistrome-higlass-wrapper#173]
The performance of horizontal multivec heatmaps highly degrades when it renders a large number of rows (1,000~).

Possible options to address this issue would be aggregating values of adjacent rows from the higlass server by considering the number of rows and/or the size of screen space. 

We also need to look into how the tile sizes are determined for the multivec heatmaps.

**Comment by sehilyi** [comment]
We can think of using tileset concepts along the y axis. However, because the arrangement of rows along y-axis is not fixed, they do not look feasible in our case.

**Comment by keller-mark** [comment]
I agree that tiling along the y axis would be difficult or would restrict us to a fixed arrangement of rows. However, we should still look into the following comment that was made on the google doc, related to the performance challenges:
> what's the bottleneck? Loading the data, rendering, or handling the state?

**Comment by sehilyi** [comment]
Yes, I think that is a good point!

I think, for the `data` part, we can print time points before and after loading data, possibly in the source code of `higlass server`?

And for the `rendering` part, I don't have better idea than printing a time point right before the rendering and manually determining time of done rendering.

If you have any better idea, please let me know.

And do you have any idea what is the the "state" in that sentence?

**Comment by sehilyi** [comment]
Related questions/comments we have after the monthly meeting today:

> We should test the performance (1) in terms of the varying number of rows in a horizontal-multivec track and (2) the varying number of BigWig files for multiple horizontal trasks.
> - What does it mean to be slow?
>    - Amount of memory used, browser crash, time to render, time to load?
>    - Assume Macbook Pro with a fixed screen resolution
>      - (Since larger screen resolution could mean that more tiles are requested)
> - Make calculations related to the scalability as a function of the number or size of the bigwig file inputs
>    - How much storage would a server need to have in order to store all cistrome files after they have been converted to multivec files?
>       - Does it scale linearly?
>   - How much time would this take?

**Comment by sehilyi** [comment]
I performed a performance test using the varying number of rows (200, 500, 1000, 5000, 10000) and view sizes (1742x938, 875x488):

https://github.com/sehilyi/higlass-horizontal-multivec-performance

We can discuss about the results at the weekly meeting tomorrow.

**Comment by keller-mark** [comment]
Wow this is awesome. I also started to use this [lighthouse tool](https://github.com/GoogleChrome/lighthouse/blob/master/docs/readme.md#using-programmatically) but your approach of computing the times in the higlass internal functions gives much more control over the measurements

**Comment by sehilyi** [comment]
After searching for useful tools to test the peformance, I just came up with implementing my own custom codes mainly because of the flexibility. I haven't looked into the lighthouse tool in detail, but it would be really great if I can find a tool that shows all the function call stacks/history with detailed time info, and even same such data as a csv file?
